### PR TITLE
Ignore results annotated with '# noqa'

### PR DESCRIPTION
**Pull request by RJ722**
Hey, 

## Description

This PR isn't production-ready by any means, but I just wanted to have an idea about the complexity it would add to Vulture to support `# noqa` comments. I'd be happy to add tests and documentation, once we all settle if it would be a good idea to carry this forward.

Another change with the introduction of `# noqa` comments would be the introduction of issue codes (it's possible without these as well, but would hurt specificity). For now, I've just added `V001`, `V002`, ... - the nomenclature for these issue codes is another thing to think about.

## Related Issue

#160, #182, #190 

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation in the README file.
- [x] I have updated the documentation accordingly.
- [x] I have added an entry in CHANGELOG.md.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: vulture/core.py [hunk 1]
```diff
@@ -46,6 +47,39 @@
...
+    # `# noqa:E123,W451,F921`
+    # `# NoQA: E123,W451,F921`
+    # `# NOQA: E123,W451,F921`
+    # `# NOQA:E123,W451,F921`
+    # We do not want to capture the ``: `` that follows ``noqa``
+    # We do not care about the casing of ``noqa``
+    # We want a comma-separated list of errors
+    # https://regex101.com/r/4XUuax/3 full explenation of the regex
+    r"#( )?noqa(?::[\s]?(?P<codes>([A-Z]+[0-9]+(?:[,\s]+)?)+))?",
+    re.IGNORECASE,
```
Comment (RJ722) [inline vulture/core.py]: This expression was taken from https://gitlab.com/pycqa/flake8//blob/0ac337608b3d88079aa91bdcffed7b4df6fd28a8/src/flake8/defaults.py#L26 with an added case of `#noqa` (without a space b/w `#` and `noqa`) -- explanation URL updated.

Comment (jendrikseipp) [reply vulture/core.py]: I think it makes sense to follow exactly what flake8 uses.

File: vulture/core.py [hunk 2]
```diff
@@ -129,6 +164,7 @@ def __init__(
         self.last_lineno = last_lineno
         self.message = message or "unused {typ} '{name}'".format(**locals())
         self.confidence = confidence
+        self.issue_code = ISSUE_CODE_MAP[typ]
```
Comment (jendrikseipp) [inline vulture/core.py]: No need to store this here if it's easily deducible from the `typ`.

File: vulture/core.py [hunk 3]
```diff
@@ -201,14 +237,27 @@ def get_set(typ):
...
         self.ignore_names = ignore_names or []
         self.ignore_decorators = ignore_decorators or []
+        self.noqa_matches = defaultdict(list)
 
         self.filename = ""
         self.code = []
         self.found_dead_code_or_error = False
 
+    def add_noqa_match(self, code, lineno):
+        self.noqa_matches[code].append(lineno)
```
Comment (jendrikseipp) [inline vulture/core.py]: No separate method needed.

File: vulture/core.py [hunk 4]
```diff
@@ -46,6 +47,39 @@
...
+    # We do not want to capture the ``: `` that follows ``noqa``
+    # We do not care about the casing of ``noqa``
+    # We want a comma-separated list of errors
+    # https://regex101.com/r/4XUuax/3 full explenation of the regex
+    r"#( )?noqa(?::[\s]?(?P<codes>([A-Z]+[0-9]+(?:[,\s]+)?)+))?",
+    re.IGNORECASE,
+)
+
+
+def _find_noqa(line):
```
Comment (jendrikseipp) [inline vulture/core.py]: No separate method needed.

File: vulture/core.py [hunk 5]
```diff
@@ -201,14 +237,27 @@ def get_set(typ):
...
 
         self.filename = ""
         self.code = []
         self.found_dead_code_or_error = False
 
+    def add_noqa_match(self, code, lineno):
+        self.noqa_matches[code].append(lineno)
+
+    def parse_noqa(self, code):
+        for index, line in enumerate(code):
```
Comment (jendrikseipp) [inline vulture/core.py]: enumerate(code, start=1)

File: vulture/core.py [hunk 6]
```diff
@@ -201,14 +237,27 @@ def get_set(typ):
...
+            match = _find_noqa(line)
+            if match is not None:
```
Comment (jendrikseipp) [inline vulture/core.py]: `if match`

File: vulture/core.py [hunk 7]
```diff
@@ -201,14 +237,27 @@ def get_set(typ):
...
+                code = match.groupdict()["codes"] or "all"
```
Comment (jendrikseipp) [inline vulture/core.py]: This needs to handle the case when multiple codes are given. (split by commas and strip whitespace, I guess.)

File: vulture/core.py [hunk 8]
```diff
@@ -282,6 +331,16 @@ def get_unused_code(self, min_confidence=0, sort_by_size=False):
...
         Return ordered list of unused Item objects.
         """
+
+        def _has_noqa(obj):
+            code = (
+                obj.issue_code
+                if obj.issue_code in self.noqa_matches
+                else "all"
+            )
+            if code in self.noqa_matches:
```
Comment (jendrikseipp) [inline vulture/core.py]: I think this line (the if-test) can be removed.

File: vulture/core.py [hunk 9]
```diff
@@ -201,14 +237,27 @@ def get_set(typ):
 
         self.ignore_names = ignore_names or []
         self.ignore_decorators = ignore_decorators or []
+        self.noqa_matches = defaultdict(list)
```
Comment (jendrikseipp) [inline vulture/core.py]: This should be a set instead of a list.

File: vulture/core.py [hunk 10]
```diff
@@ -301,8 +360,14 @@ def by_size(item):
             + self.unreachable_code
         )
 
+        results_without_noqa = [
+            obj for obj in unused_code if not _has_noqa(obj)
+        ]
```
Comment (jendrikseipp) [inline vulture/core.py]: No need to create a filtered list. Simply filter objects below.

File: vulture/core.py [hunk 11]
```diff
@@ -282,6 +331,16 @@ def get_unused_code(self, min_confidence=0, sort_by_size=False):
...
         """
+
+        def _has_noqa(obj):
+            code = (
+                obj.issue_code
+                if obj.issue_code in self.noqa_matches
+                else "all"
+            )
+            if code in self.noqa_matches:
+                return obj.first_lineno in self.noqa_matches[code]
```
Comment (jendrikseipp) [inline vulture/core.py]: I think the whole function can be reduced to `return obj.first_lineno in self.noqa_matches["all"] or obj.first_lineno in self.noqa_matches[ISSUE_CODE_MAP[obj.typ]]`

File: vulture/core.py (hunk 1 above)
Comment (RJ722) [reply vulture/core.py]: Okay. Changed it back.

**Comment by RJ722** [comment]
@jendrikseipp should `ISSUE_CODE_MAP` be a class variable for `Vulture`?

I'm asking because my guess is that most folks would only import Vulture, if using the API (`from vulture import Vulture`).

File: vulture/core.py (hunk 11 above)
Comment (RJ722) [reply vulture/core.py]: I ended up with:

```py
return obj.first_lineno in (
    self.noqa_matches[ISSUE_CODE_MAP[obj.typ]]
    or self.noqa_matches["all"]
)
```

File: vulture/core.py (hunk 10 above)
Comment (RJ722) [reply vulture/core.py]: Yeah, that improves readability. thanks!

File: vulture/core.py (hunk 7 above)
Comment (RJ722) [reply vulture/core.py]: Done!

File: vulture/core.py (hunk 5 above)
Comment (RJ722) [reply vulture/core.py]: thanks!

File: vulture/core.py [hunk 12]
```diff
@@ -46,6 +47,34 @@
 if sys.version_info < (3, 4):
     IGNORED_VARIABLE_NAMES |= {"True", "False"}
 
+ISSUE_CODE_MAP = {
```
Comment (jendrikseipp) [inline vulture/core.py]: Let's call this dict ERROR_CODES. I think leaving this as a global variable is fine.

File: vulture/core.py [hunk 13]
```diff
@@ -282,6 +325,13 @@ def get_unused_code(self, min_confidence=0, sort_by_size=False):
         """
         Return ordered list of unused Item objects.
         """
+
+        def _has_noqa(obj):
+            return obj.first_lineno in (
+                self.noqa_matches[ISSUE_CODE_MAP[obj.typ]]
+                or self.noqa_matches["all"]
+            )
```
Comment (jendrikseipp) [inline vulture/core.py]: This will fail if, e.g., "noqa V001" is given for line 10, "noqa" is given for line 5 and the object appears on line 5.

File: vulture/core.py [hunk 14]
```diff
@@ -282,6 +325,13 @@ def get_unused_code(self, min_confidence=0, sort_by_size=False):
         """
         Return ordered list of unused Item objects.
         """
+
+        def _has_noqa(obj):
```
Comment (jendrikseipp) [inline vulture/core.py]: No reason for leading underscore in nested function.

File: vulture/core.py [hunk 15]
```diff
@@ -46,6 +47,34 @@
...
+    # `# noqa: E123`
+    # `# noqa: E123,W451,F921`
+    # `# noqa:E123,W451,F921`
+    # `# NoQA: E123,W451,F921`
+    # `# NOQA: E123,W451,F921`
+    # `# NOQA:E123,W451,F921`
+    # We do not want to capture the ``: `` that follows ``noqa``
+    # We do not care about the casing of ``noqa``
+    # We want a comma-separated list of errors
+    # https://regex101.com/r/4XUuax/2 full explanation of the regex
```
Comment (jendrikseipp) [inline vulture/core.py]: No need to copy the whole comment. The first three lines and a well-formed example are enough.

File: vulture/core.py [hunk 16]
```diff
@@ -201,14 +230,28 @@ def get_set(typ):
...
         self.found_dead_code_or_error = False
 
+    def parse_noqa(self, code):
+        for lineno, line in enumerate(code, start=1):
+            match = NOQA_REGEXP.search(line)
+            if match:
+                codes = [
+                    c.strip()
+                    for c in (match.groupdict()["codes"] or "all").split(",")
+                    # if no code is specified, assign it to `all`
```
Comment (jendrikseipp) [inline vulture/core.py]: Comment above code.

File: vulture/core.py [hunk 17]
```diff
@@ -301,8 +351,11 @@ def by_size(item):
             + self.unreachable_code
         )
 
+        # breakpoint()
         confidently_unused = [
```
Comment (jendrikseipp) [inline vulture/core.py]: Rename to "reported_unused".

File: vulture/core.py (hunk 13 above)
Comment (RJ722) [reply vulture/core.py]: Oh yes, nice catch. We need a union of two sets here.

**Comment by RJ722** [comment]
Another thing I was pondering about. We don't start ignoring `noqa` until `get_unused_code` is called. What if the user bypasses it, e.g.:

```py
from vulture import Vulture

v = Vulture()
v.scan('import this  # noqa: V004')
print(v.unused_imports)
```

Should we move the `has_noqa` logic to `_define` method?

**Comment by jendrikseipp** [comment]
> Should we move the `has_noqa` logic to `_define` method?

Good idea! This aligns better with the existing code for ignoring things.

**Comment by RJ722** [comment]
TODO:

- [ ] One test is failing because Python3.8 onwards, the first line for property node is the same line at which the function starts.
- [ ] Write documentation for these changes -- @jendrikseipp can you take this on, please? 😅

File: tests/test_noqa.py [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+"""
+    )
+    check(v.unused_imports, ["this"])
+
+
+def test_noqa_with_special_unicode(v):
+    v.scan(
+        """\
+import abc  # noqa: V012, V😎12
+import problems  # noqa: V03🙃1
```
Comment (RJ722) [inline tests/test_noqa.py]: This test is the culprit behind tests failing for Python2.7. Should I remove them?

Comment (jendrikseipp) [reply tests/test_noqa.py]: Yes, I think it's easiest to remove the emojis.

**Comment by jendrikseipp** [comment]
Could you also add a test that verifies that we fixed the "or vs. union" bug?

**Comment by RJ722** [comment]
> Could you also add a test that verifies that we fixed the "or vs. union" bug?

The way I've added the tests is basically ensuring that an example of the same issue code is  covered in both categories - `all` and `V00X` (on different lines - both before and after). If there'd have been a problem with "or v/s union", we'd have seen it by now.

**Comment by RJ722** [comment]
But, now that I come think of it, I haven't added isolated cases - just `V00X`, or `# noqa` alone. Adding them.

File: vulture/core.py [hunk 19]
```diff
@@ -457,26 +495,41 @@ def _define(
...
+        def ignored(name):
+            return (ignore and ignore(self.filename, name)) or _match(
+                name, self.ignore_names
+            )
+
+        def has_noqa(lineno, typ):
+            if typ == "property" and sys.version_info >= (3, 8):
+                # python3.8 onwards, lineno for property is the same as that of
+                # the decorated method.
+                lineno = lineno - 1
```
Comment (RJ722) [inline vulture/core.py]: Also, I think, it rather makes sense to ignore this case:

```py
class Foo:
    @property
    def bar():  # noqa
        pass
```

than adding `# noqa` infront of `@property`?

File: tests/test_noqa.py [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+    match = NOQA_REGEXP.search(line)
+    parsed = [
+        c.strip() for c in (match.groupdict()["codes"] or "all").split(",")
+    ]
+    assert parsed == codes
+
+
+@pytest.mark.parametrize(
+    "line",
+    [("# noqa: 123V"), ("# noqa explaination: V012"), ("# noqa: ,V001"),],
```
Comment (RJ722) [inline tests/test_noqa.py]: Remove trailing comma.

File: tests/test_noqa.py [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+        ("# noqa Hi, go on.", ["all"]),
+        ("# noqa: V001", ["V001"]),
+        ("# noqa: V001, V007", ["V001", "V007"]),
+        ("# NoQA: V001,      V003, \t V004", ["V001", "V003", "V004"]),
+    ],
+)
+def test_noqa_regex_present(line, codes):
+    match = NOQA_REGEXP.search(line)
+    parsed = [
+        c.strip() for c in (match.groupdict()["codes"] or "all").split(",")
```
Comment (jendrikseipp) [inline tests/test_noqa.py]: Can you factor this out into a function in core.py?

File: tests/test_report.py [hunk 22]
```diff
@@ -39,13 +39,13 @@ def test_report(code, expected, make_whitelist=False):
...
-{filename}:9: unused variable 'foobar' (60% confidence)
-{filename}:11: unreachable code after 'return' (100% confidence)
-{filename}:13: unused property 'myprop' (60% confidence)
+<V004> {filename}:1: unused import 'foo' (90% confidence)
+<V002> {filename}:3: unused class 'Foo' (60% confidence)
+<V003> {filename}:7: unused function 'bar' (60% confidence)
+<V001> {filename}:8: unused attribute 'foobar' (60% confidence)
+<V007> {filename}:9: unused variable 'foobar' (60% confidence)
+<V006> {filename}:11: unreachable code after 'return' (100% confidence)
+<V005> {filename}:13: unused property 'myprop' (60% confidence)
```
Comment (jendrikseipp) [inline tests/test_report.py]: I think we should mimic the flake8 format more closely (test.py:2:5: F841 local variable 'foo' is assigned to but never used), so for example:

{filename}:1: V004 unused import 'foo' (90% confidence)

File: tests/test_noqa.py [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+
+class Cellar:  # noqa: V002
+    @property  # noqa: V005
+    def wine(self, grapes):  # noqa: V007
+        pass
+
+    def serve(self, quantity=50):  # noqa: V003
+        self.quantity_served = quantity  # noqa: V001
+        return
+        self.pour()  # noqa: V006
```
Comment (jendrikseipp) [inline tests/test_noqa.py]: Shouldn't this be V007?

File: vulture/core.py (hunk 19 above)
Comment (jendrikseipp) [reply vulture/core.py]: I think adding the noqa on the same line as bar() is the better solution, yes. We should change this at a lower level as well: in all Python versions, we should report the line containing bar() as the line number for the property. This should be a separate PR though. However, this might be tricky in case there are multiple @property decorators. Could you try this out, please?

File: vulture/core.py [hunk 24]
```diff
@@ -457,26 +495,41 @@ def _define(
...
+                )
+            )
+
         last_node = last_node or first_node
         typ = collection.typ
-        if (ignore and ignore(self.filename, name)) or _match(
-            name, self.ignore_names
-        ):
+        first_lineno = first_node.lineno
+        if ignored(name) or has_noqa(first_lineno, typ):
```
Comment (jendrikseipp) [inline vulture/core.py]: `has_noqa` can become part of `ignored`

File: vulture/core.py [hunk 25]
```diff
@@ -46,6 +47,28 @@
...
 
+ERROR_CODES = {
+    "attribute": "V001",
+    "class": "V002",
+    "function": "V003",
+    "import": "V004",
+    "property": "V005",
+    "unreachable_code": "V006",
+    "variable": "V007",
+}
```
Comment (jendrikseipp) [inline vulture/core.py]: I think we should follow pycodestyle and start with error code V101 instead of V001. Since unreachable_code is somewhat different from unused code, I'd give it number V201 (and unused variables V107). (I see now that you changed the order.)

Comment (RJ722) [reply vulture/core.py]: If we start with V101, and don't assign unreachable_code in that series, unused_variables would be V106, no?

Comment (jendrikseipp) [reply vulture/core.py]: That's right :-)

File: tests/test_noqa.py [hunk 26]
```diff
@@ -0,0 +1,274 @@
...
+def test_noqa_regex_no_groups(line):
+    assert NOQA_REGEXP.search(line).groupdict()["codes"] is None
+
+
+@pytest.mark.parametrize(
+    "line",
+    [("#noqa"), ("##noqa"), ("# n o q a"), ("#NOQA"), ("# Hello, noqa")],
+)
+def test_noqa_regex_not_present(line):
+    assert bool(NOQA_REGEXP.search(line)) is False
```
Comment (jendrikseipp) [inline tests/test_noqa.py]: Why not `assert not NOQA_REGEXP.search(line)`?

File: tests/test_noqa.py [hunk 27]
```diff
@@ -0,0 +1,274 @@
...
+    match = NOQA_REGEXP.search(line)
+    parsed = _parse_error_codes(match)
+    assert parsed == codes
+
+
+@pytest.mark.parametrize(
+    "line",
+    [
+        ("# noqa: 123V"),
+        ("# noqa explaination: V012"),
```
Comment (jendrikseipp) [inline tests/test_noqa.py]: explaination -> explanation

File: tests/test_noqa.py [hunk 28]
```diff
@@ -0,0 +1,274 @@
...
+        self.pour()  # noqa
+"""
+    )
+    check(v.unused_attrs, [])
+    check(v.unused_classes, [])
+    check(v.unused_funcs, [])
+    check(v.unused_imports, [])
+    check(v.unused_props, [])
+    check(v.unreachable_code, [])
+    check(v.unused_vars, [])
```
Comment (jendrikseipp) [inline tests/test_noqa.py]: Can you use exactly one noqa for each category?

File: tests/test_noqa.py [hunk 29]
```diff
@@ -0,0 +1,274 @@
...
+
+
+def test_noqa_specific_issue_codes(v):
+    v.scan(
+        """\
+import this  # noqa: V104
+
+class Cellar:  # noqa: V102
+    @property
+    def wine(self, grapes):  # noqa: V105, V106
```
Comment (jendrikseipp) [inline tests/test_noqa.py]: Same here. let's add a separate unused variable inside of this method. The we can also remove the `pass`.

File: tests/test_noqa.py [hunk 30]
```diff
@@ -0,0 +1,274 @@
...
+def test_noqa_functions(v):
+    v.scan(
+        """\
+def play(tune, instrument='bongs', _hz='50'):  # noqa: V103
+    pass
+
+
+# noqa
+def problems():  # noqa: V104
+    ''' They don't go away. :-)'''
```
Comment (jendrikseipp) [inline tests/test_noqa.py]: Too pessimistic for my taste :-) I'd remove the comment.

File: tests/test_noqa.py [hunk 31]
```diff
@@ -0,0 +1,274 @@
...
+            a_sheep.grow_hair()  # noqa: V201
+        a_sheep.shave()
+    return
+    for a_sheep in sheep:  # noqa: V201
+        if a_sheep.still_has_hair:
+            a_sheep.shave_again()
+"""
+    )
+    check(v.unreachable_code, [])
+    check(v.unused_funcs, ["shave_sheep"])
```
Comment (jendrikseipp) [inline tests/test_noqa.py]: Hehe, nice code :-)

File: tests/test_report.py [hunk 32]
```diff
@@ -39,13 +39,13 @@ def test_report(code, expected, make_whitelist=False):
...
-{filename}:9: unused variable 'foobar' (60% confidence)
-{filename}:11: unreachable code after 'return' (100% confidence)
-{filename}:13: unused property 'myprop' (60% confidence)
+{filename}:1: V104 unused import 'foo' (90% confidence)
+{filename}:3: V102 unused class 'Foo' (60% confidence)
+{filename}:7: V103 unused function 'bar' (60% confidence)
+{filename}:8: V101 unused attribute 'foobar' (60% confidence)
+{filename}:9: V106 unused variable 'foobar' (60% confidence)
+{filename}:11: V201 unreachable code after 'return' (100% confidence)
+{filename}:13: V105 unused property 'myprop' (60% confidence)
```
Comment (jendrikseipp) [inline tests/test_report.py]: Can you add to the changelog that the output format changes?

File: vulture/core.py [hunk 33]
```diff
@@ -461,26 +501,45 @@ def _define(
...
+            # if name not ignored, check if annotated with "# noqa"
+            if typ == "property" and sys.version_info < (3, 8):
+                # for python < 3.8, increment as many line numbers as there are
+                # decorators on the method.
+                # python3.8 onwards, lineno for property is the same as that of
+                # the decorated method.
+                lineno = lineno + len(first_node.decorator_list)
+            return lineno in (
+                self.noqa_matches[ERROR_CODES[typ]].union(
+                    self.noqa_matches["all"]
```
Comment (jendrikseipp) [inline vulture/core.py]: I think it's faster and more readable to test the two sets separately instead of computing the union just for the test.

File: vulture/core.py [hunk 34]
```diff
@@ -461,26 +501,45 @@ def _define(
...
+            ):
+                return True
+
+            # if name not ignored, check if annotated with "# noqa"
+            if typ == "property" and sys.version_info < (3, 8):
+                # for python < 3.8, increment as many line numbers as there are
+                # decorators on the method.
+                # python3.8 onwards, lineno for property is the same as that of
+                # the decorated method.
+                lineno = lineno + len(first_node.decorator_list)
```
Comment (jendrikseipp) [inline vulture/core.py]: I think this means that the line numbers we report for unused properties now differ from the line numbers that must have noqa on them. This should be fixed in a separate PR. Can you add a TODO to the TODO file, please?

File: vulture/core.py [hunk 35]
```diff
@@ -461,26 +501,45 @@ def _define(
...
+                name, self.ignore_names
+            ):
+                return True
+
+            # if name not ignored, check if annotated with "# noqa"
+            if typ == "property" and sys.version_info < (3, 8):
+                # for python < 3.8, increment as many line numbers as there are
+                # decorators on the method.
+                # python3.8 onwards, lineno for property is the same as that of
+                # the decorated method.
```
Comment (jendrikseipp) [inline vulture/core.py]: Please start the first sentence with an uppercase "For" and the second with "From".

File: tests/test_noqa.py (hunk 26 above)
Comment (RJ722) [reply tests/test_noqa.py]: I just didn't happen to think of it. ':D Changing it.

File: tests/test_noqa.py (hunk 28 above)
Comment (RJ722) [reply tests/test_noqa.py]: Do you mean, break it down into individual tests for all of them?

File: vulture/core.py (hunk 34 above)
Comment (RJ722) [reply vulture/core.py]: I raised #200 :D -- I would be happy to rebase this branch, once we get it merged.

File: vulture/core.py (hunk 33 above)
Comment (RJ722) [reply vulture/core.py]: That might result in the same or v/s union bug, I guess. Do you have something in mind?

File: tests/test_noqa.py (hunk 29 above)
Comment (RJ722) [reply tests/test_noqa.py]: I added the variable - `year`. Just making sure that you in fact, meant this. If yes, can we mark this as resolved?

File: tests/test_noqa.py (hunk 28 above)
Comment (jendrikseipp) [reply tests/test_noqa.py]: No, I was just surprised to see 6 noqas but 7 checks. Adding the `year` variable here as well should resolve this.

File: tests/test_noqa.py (hunk 29 above)
Comment (jendrikseipp) [reply tests/test_noqa.py]: I'd drop the useless return and remove the V106 noqa from the line above, but then it's resolved, yes.

File: vulture/core.py (hunk 33 above)
Comment (jendrikseipp) [reply vulture/core.py]: `return lineno in self.noqa_matches[ERROR_CODES[typ]] or lineno in self.noqa_matches["all"]`

Comment (RJ722) [reply vulture/core.py]: Yep, thanks!

File: tests/test_noqa.py (hunk 29 above)
Comment (RJ722) [reply tests/test_noqa.py]: Ahh, got it.

File: tests/test_noqa.py (hunk 28 above)
Comment (RJ722) [reply tests/test_noqa.py]: Aah, okay. I took care of this.

File: CHANGELOG.md [hunk 36]
```diff
@@ -1,6 +1,9 @@
 # Unreleased
 
 * Ignore unused import statements which occur in `__init__.py` (RJ722, #192).
+* Report issue codes in ouput (e.g., `code.py:1: V104 unused import ...`)
```
Comment (jendrikseipp) [inline CHANGELOG.md]: ouput -> output

File: vulture/core.py [hunk 37]
```diff
@@ -55,6 +78,13 @@ def _get_unused_items(defined_items, used_names):
     return unused_items
 
 
+def _parse_error_codes(match):
+    # if no code is specified, assign it to `all`
```
Comment (jendrikseipp) [inline vulture/core.py]: Start with upper case and end with full stop. I#d also be a little more concrete: 

If no error code is specified, add the line to the "all" category.

File: vulture/core.py [hunk 38]
```diff
@@ -461,26 +501,44 @@ def _define(
         confidence=DEFAULT_CONFIDENCE,
         ignore=None,
     ):
+        def ignored(lineno):
+            if (ignore and ignore(self.filename, name)) or _match(
+                name, self.ignore_names
+            ):
+                return True
+
+            # if name not ignored, check if annotated with "# noqa"
```
Comment (jendrikseipp) [inline vulture/core.py]: Rephrase: Check if the reported line is annotated with "# noqa".

**Comment by RJ722** [comment]
Now, that #200 has been merged (yay! \o/), let me rebase this branch and resolve conflicts.

File: CHANGELOG.md [hunk 39]
```diff
@@ -3,6 +3,9 @@
 * Ignore unused import statements which occur in `__init__.py` (RJ722, #192).
 * Report first decorator's line number for decorated objects on Python 3.8+.
   (RJ722, #200)
+* Report issue codes in output (e.g., `code.py:1: V104 unused import ...`)
+  (RJ722, #195)
+* Support `# noqa` comments to suppress results on that line. (RJ722, #195).
```
Comment (jendrikseipp) [inline CHANGELOG.md]: Could you please draft the documentation for the README.md file? It's always easier to edit existing text than writing new text.

File: vulture/core.py [hunk 40]
```diff
@@ -205,14 +236,23 @@ def get_set(typ):
 
         self.ignore_names = ignore_names or []
         self.ignore_decorators = ignore_decorators or []
+        self.noqa_matches = defaultdict(set)
 
         self.filename = ""
         self.code = []
         self.found_dead_code_or_error = False
 
+    def parse_noqa(self, code):
```
Comment (jendrikseipp) [inline vulture/core.py]: It's a bit strange that the method is passed `self.code`, but not `self.noqa_matches` even though both are instance attributes. I think we should pass both variables to the method and turn it into a function. Then we can collect all code related to noqa in a vulture/noqa.py file, i.e., parse_noqa(), _parse_error_codes() and NOQA_REGEXP.

File: vulture/core.py [hunk 41]
```diff
@@ -461,14 +501,25 @@ def _define(
...
+            if (ignore and ignore(self.filename, name)) or _match(
+                name, self.ignore_names
+            ):
+                return True
+
+            # Check if the reported line is annotated with "# noqa".
+            return (
+                lineno in self.noqa_matches[ERROR_CODES[typ]]
+                or lineno in self.noqa_matches["all"]
+            )
```
Comment (jendrikseipp) [inline vulture/core.py]: This can become a function `noqa.ignore_line(noqa_matches, lineno, typ)` and then we can add it to the if-statement above.

File: vulture/core.py [hunk 42]
```diff
@@ -205,14 +236,23 @@ def get_set(typ):
 
         self.ignore_names = ignore_names or []
         self.ignore_decorators = ignore_decorators or []
+        self.noqa_matches = defaultdict(set)
```
Comment (jendrikseipp) [inline vulture/core.py]: Please rename noqa_matches to noqa_lines everywhere.

File: CHANGELOG.md (hunk 39 above)
Comment (RJ722) [reply CHANGELOG.md]: Sure!

**Comment by jendrikseipp** [comment]
Thanks for the additional changes! I edited README.md slightly. Could you please resolve the merge conflicts? Then I can merge the PR.

**Comment by RJ722** [comment]
@jendrikseipp, thanks for your timely reviews, and help on this! :-)
